src/transforms/groupby.js

```diff
@@ -82,44 +90,85 @@ exports.transform = function(data, state) {
 
     var newData = [];
 
-    data.forEach(function(trace) {
-        newData = newData.concat(transformOne(trace, state));
+    data.forEach(function(trace, i) {
+
+        newData = newData.concat(transformOne(trace, state, state.attributeSets[i]));
     });
 
     return newData;
 };
 
-function transformOne(trace, state) {
+function transformOne(trace, state, attributeSet) {
+
     var opts = state.transform;
-    var groups = opts.groups;
+    var groups = trace.transforms[state.transformIndex].groups;
 
     var groupNames = groups.filter(function(g, i, self) {
         return self.indexOf(g) === i;
     });
 
     var newData = new Array(groupNames.length);
-    var len = Math.min(trace.x.length, trace.y.length, groups.length);
+    var len = groups.length;
+
+    var style = opts.style || {};
 
+    var topLevelAttributes = attributeSet
+        .filter(function(array) {return Array.isArray(getDeepProp(trace, array));});
+
+    var initializeArray = function(newTrace, a) {
+        setDeepProp(newTrace, a, []);
+    };
+
+    var pasteArray = function(newTrace, trace, j, a) {
+        getDeepProp(newTrace, a).push(getDeepProp(trace, a)[j]);
+    };
+
+    // fixme the O(n**3) complexity
     for(var i = 0; i < groupNames.length; i++) {
         var groupName = groupNames[i];
 
         // TODO is this the best pattern ???
         // maybe we could abstract this out
         var newTrace = newData[i] = Lib.extendDeep({}, trace);
 
-        newTrace.x = [];
-        newTrace.y = [];
+        topLevelAttributes.forEach(initializeArray.bind(null, newTrace));
 
         for(var j = 0; j < len; j++) {
             if(groups[j] !== groupName) continue;
 
-            newTrace.x.push(trace.x[j]);
-            newTrace.y.push(trace.y[j]);
+            topLevelAttributes.forEach(pasteArray.bind(0, newTrace, trace, j));
         }
 
         newTrace.name = groupName;
-        newTrace.marker.color = opts.groupColors[groupName];
+
+        //  there's no need to coerce style[groupName] here
+        // as another round of supplyDefaults is done on the transformed traces
+        newTrace = Lib.extendDeep(newTrace, style[groupName] || {});
     }
 
     return newData;
 }
+
+function getDeepProp(thing, propArray) {
```

Comment:
`getDeepProp` and `setDeepProp` would be nice additions to [`src/lib/nested_property.js`](https://github.com/plotly/plotly.js/blob/master/src/lib/nested_property.js).